shared-bindings/dotclockframebuffer/DotClockFramebuffer.c

```diff
@@ -0,0 +1,338 @@
+        { MP_QSTR_vsync_front_porch, MP_ARG_INT | MP_ARG_KW_ONLY | MP_ARG_REQUIRED, {.u_int = 0 } },
+        { MP_QSTR_vsync_idle_low, MP_ARG_BOOL | MP_ARG_KW_ONLY | MP_ARG_REQUIRED, {.u_bool = false } },
+
+        { MP_QSTR_de_idle_high, MP_ARG_BOOL | MP_ARG_KW_ONLY | MP_ARG_REQUIRED, {.u_bool = false } },
+        { MP_QSTR_pclk_active_high, MP_ARG_BOOL | MP_ARG_KW_ONLY | MP_ARG_REQUIRED, {.u_bool = false } },
+        { MP_QSTR_pclk_idle_high, MP_ARG_BOOL | MP_ARG_KW_ONLY | MP_ARG_REQUIRED, {.u_bool = false } },
+    };
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    const mcu_pin_obj_t *de = validate_obj_is_free_pin(args[ARG_de].u_obj, MP_QSTR_de);
+    const mcu_pin_obj_t *vsync = validate_obj_is_free_pin(args[ARG_vsync].u_obj, MP_QSTR_vsync);
+    const mcu_pin_obj_t *hsync = validate_obj_is_free_pin(args[ARG_hsync].u_obj, MP_QSTR_hsync);
+    const mcu_pin_obj_t *dclk = validate_obj_is_free_pin(args[ARG_dclk].u_obj, MP_QSTR_dclk);
+
+    uint8_t num_red, num_green, num_blue;
+    const mcu_pin_obj_t *red_pins[8], *green_pins[8], *blue_pins[8];
+
+    validate_list_is_free_pins(MP_QSTR_red, red_pins, (mp_int_t)MP_ARRAY_SIZE(red_pins), args[ARG_red].u_obj, &num_red);
+    validate_list_is_free_pins(MP_QSTR_green, green_pins, (mp_int_t)MP_ARRAY_SIZE(green_pins), args[ARG_green].u_obj, &num_green);
+    validate_list_is_free_pins(MP_QSTR_blue, blue_pins, (mp_int_t)MP_ARRAY_SIZE(blue_pins), args[ARG_blue].u_obj, &num_blue);
+
+    mp_printf(&mp_plat_print, "#red=%d #green=%d #blue=%d\n", num_red, num_green, num_blue);
+
+    dotclockframebuffer_framebuffer_obj_t *self = &allocate_display_bus_or_raise()->dotclock;
+    self->base.type = &dotclockframebuffer_framebuffer_type;
+
+    common_hal_dotclockframebuffer_framebuffer_construct(
+        self, de, vsync, hsync, dclk, red_pins, num_red, green_pins, num_green, blue_pins, num_blue,
+        args[ARG_frequency].u_int, args[ARG_width].u_int, args[ARG_height].u_int,
+        args[ARG_hsync_pulse_width].u_int, args[ARG_hsync_back_porch].u_int, args[ARG_hsync_front_porch].u_int, args[ARG_hsync_idle_low].u_bool,
+        args[ARG_vsync_pulse_width].u_int, args[ARG_vsync_back_porch].u_int, args[ARG_vsync_front_porch].u_int, args[ARG_vsync_idle_low].u_bool,
+        args[ARG_de_idle_high].u_bool,
+        args[ARG_pclk_active_high].u_bool,
+        args[ARG_pclk_idle_high].u_bool
+        );
+
+    return self;
+}
+
+static void check_for_deinit(dotclockframebuffer_framebuffer_obj_t *self) {
+    if (common_hal_dotclockframebuffer_framebuffer_deinitialized(self)) {
+        raise_deinited_error();
+    }
+}
+
+//|     def refresh(self) -> None:
+//|         """Transmits the color data in the buffer to the pixels so that
+//|         they are shown.
+//|
+//|         If this function is not called, the results are unpredictable; updates may be partially shown.
```

Comment:
Do these displays not auto-refresh by default?

Reply:
The wrinkle is that esp32 requires a cache flush. otherwise a mix of old and new data is displayed.

Reply:
Can't we do that automatically?

Reply:
This also matches rgbmatrix (https://docs.circuitpython.org/en/latest/shared-bindings/rgbmatrix/index.html#rgbmatrix.RGBMatrix.refresh) and will need to be explicit if and when we add double buffering.

this only affects folks who are accessing the raw framebuffer and does not affect displayio usage.